File: bindings/rust/src/safe_unit.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

//! Safe physical unit conversions via libproven FFI.
//!
//! Converts between length, temperature, and other physical units.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Length unit.
pub use crate::ffi::LengthUnit;

/// Temperature unit.
pub use crate::ffi::TempUnit as TemperatureUnit;

// Additional unit types for API compatibility (not all have FFI backing yet).

/// Mass unit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MassUnit {
    /// Kilograms.
    Kilograms,
    /// Grams.
    Grams,
    /// Milligrams.
    Milligrams,
    /// Pounds.
    Pounds,
    /// Ounces.
    Ounces,
}

/// Time unit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TimeUnit {
    /// Seconds.
    Seconds,
    /// Milliseconds.
    Milliseconds,
    /// Microseconds.
    Microseconds,
    /// Minutes.
    Minutes,
    /// Hours.
    Hours,
    /// Days.
    Days,
}

/// Data unit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DataUnit {
    /// Bytes.
    Bytes,
    /// Kilobytes (1000 bytes).
    Kilobytes,
    /// Megabytes.
    Megabytes,
    /// Gigabytes.
    Gigabytes,
    /// Terabytes.
    Terabytes,
    /// Kibibytes (1024 bytes).
    Kibibytes,
    /// Mebibytes.
    Mebibytes,
    /// Gibibytes.
    Gibibytes,
}

/// Convert length between units.
pub fn convert_length(value: f64, from: LengthUnit, to: LengthUnit) -> Result<f64> {
    // SAFETY: proven_unit_convert_length takes value-type arguments;
    // always safe to call.
    let result = unsafe { ffi::proven_unit_convert_length(value, from, to) };
    core::float_result_to_result(result)
}

/// Convert temperature between units.
pub fn convert_temperature(value: f64, from: TemperatureUnit, to: TemperatureUnit) -> Result<f64> {
    // SAFETY: proven_unit_convert_temp takes value-type arguments;
    // always safe to call.
    let result = unsafe { ffi::proven_unit_convert_temp(value, from, to) };
    core::float_result_to_result(result)
}

/// Convert mass between units.
///
/// Not yet available via FFI; returns `Error::NotImplemented`.
/// Will use verified Idris 2 conversion factors once the FFI export exists.
pub fn convert_mass(_value: f64, _from: MassUnit, _to: MassUnit) -> Result<f64> {
    Err(core::Error::NotImplemented)
}

/// Convert time between units.
///
/// Not yet available via FFI; returns `Error::NotImplemented`.
pub fn convert_time(_value: f64, _from: TimeUnit, _to: TimeUnit) -> Result<f64> {
    Err(core::Error::NotImplemented)
}

/// Convert data size between units.
///
/// Not yet available via FFI; returns `Error::NotImplemented`.
pub fn convert_data(_value: f64, _from: DataUnit, _to: DataUnit) -> Result<f64> {
    Err(core::Error::NotImplemented)
}

File: bindings/zig/src/safe_unit.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeUnit - FFI bindings to libproven unit conversion operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for unit operations.
pub const UnitError = error{
    InvalidArgument,
    OutOfBounds,
    ProvenError,
};

/// Length units (matching ProvenLengthUnit enum).
pub const LengthUnit = enum(c_int) {
    meters = c.PROVEN_LENGTH_METERS,
    kilometers = c.PROVEN_LENGTH_KILOMETERS,
    centimeters = c.PROVEN_LENGTH_CENTIMETERS,
    millimeters = c.PROVEN_LENGTH_MILLIMETERS,
    feet = c.PROVEN_LENGTH_FEET,
    inches = c.PROVEN_LENGTH_INCHES,
    miles = c.PROVEN_LENGTH_MILES,
    yards = c.PROVEN_LENGTH_YARDS,
};

/// Temperature units (matching ProvenTempUnit enum).
pub const TempUnit = enum(c_int) {
    celsius = c.PROVEN_TEMP_CELSIUS,
    fahrenheit = c.PROVEN_TEMP_FAHRENHEIT,
    kelvin = c.PROVEN_TEMP_KELVIN,
};

/// Convert length between units via libproven.
pub fn convertLength(value: f64, from: LengthUnit, to: LengthUnit) UnitError!f64 {
    const result = c.proven_unit_convert_length(value, @intFromEnum(from), @intFromEnum(to));
    return switch (result.status) {
        c.PROVEN_OK => result.value,
        c.PROVEN_ERR_INVALID_ARGUMENT => error.InvalidArgument,
        else => error.ProvenError,
    };
}

/// Convert temperature between units via libproven.
pub fn convertTemp(value: f64, from: TempUnit, to: TempUnit) UnitError!f64 {
    const result = c.proven_unit_convert_temp(value, @intFromEnum(from), @intFromEnum(to));
    return switch (result.status) {
        c.PROVEN_OK => result.value,
        c.PROVEN_ERR_OUT_OF_BOUNDS => error.OutOfBounds,
        else => error.ProvenError,
    };
}

test "convertLength" {
    const meters = try convertLength(1.0, .kilometers, .meters);
    try std.testing.expectApproxEqAbs(@as(f64, 1000.0), meters, 0.01);
}

test "convertTemp" {
    const fahrenheit = try convertTemp(100.0, .celsius, .fahrenheit);
    try std.testing.expectApproxEqAbs(@as(f64, 212.0), fahrenheit, 0.01);
}
